### DISCUSSION TITLE
Blending procedure on overlapping patches

### DISCUSSION THREAD

**Discussion by geografif** [opengeos/geoai#309] [Q&A] [answered]
Hi @giswqs, I have a question regarding the blending procedure on overlapping patches applied to a large image inference. Some methodological research papers and tools (e.g. scikit-eo) explicitly mentions 'averaging' was used to handle overlapping prediction patch. However, for my use case, using average didn't seem to be a good option.

If I am not mistaken, the blending script is defined under `train.py > def inference_on_geotiff`,  which are:

`
# Process batch when it reaches the batch size or at the end
                if batch_count == batch_size or (i == steps_y and j == steps_x):
                    # Forward pass
                    with torch.no_grad():
                        outputs = model(batch_inputs)

                    # Process each output in the batch
                    for idx, output in enumerate(outputs):
                        y_pos, x_pos, h, w = batch_positions[idx]

                        # Create weight matrix that gives higher weight to center pixels
                        # This helps with smooth blending at boundaries
                        y_grid, x_grid = np.mgrid[0:h, 0:w]

                        # Calculate distance from each edge
                        dist_from_left = x_grid
                        dist_from_right = w - x_grid - 1
                        dist_from_top = y_grid
                        dist_from_bottom = h - y_grid - 1

                        # Combine distances (minimum distance to any edge)
                        edge_distance = np.minimum.reduce(
                            [
                                dist_from_left,
                                dist_from_right,
                                dist_from_top,
                                dist_from_bottom,
                            ]
                        )

                        # Convert to weight (higher weight for center pixels)
                        # Normalize to [0, 1]
                        edge_distance = np.minimum(edge_distance, overlap / 2)
                        weight = edge_distance / (overlap / 2)
`

Could you please explain what procedure does `geoai` use?

Thanks in advance.

**Comment by giswqs** [comment, marked as answer]
It uses weighted blending with distance-based weights to process edge pixels:

  1. Overlapping tiles: Windows overlap by a specified amount (default 256 pixels) - line 1024
  2. Distance-based weighting: For each tile, pixels are weighted based on their distance from the nearest edge (lines 1162-1183):
    - Calculates minimum distance to any edge (top, bottom, left, right)
    - Pixels at edges get weight ~0, center pixels get weight ~1
    - Weight transitions smoothly within the overlap zone (overlap/2)
  3. Accumulation: Two arrays track predictions (lines 1071-1073):
    - pred_accumulator: Accumulates weighted predictions
    - count_accumulator: Tracks sum of weights per pixel
  4. Blending: Overlapping regions are blended by adding weighted predictions and dividing by total weights (lines 1202-1207)

  This approach gives higher confidence to center pixels of each tile and smoothly blends predictions in overlapping regions, reducing edge artifacts compared to simple averaging or hard boundaries.
